src/object/cli_shard.c

```diff
@@ -222,20 +225,40 @@ int dc_rw_cb_csum_verify(const struct rw_cb_args *rw_args)
 					     iod_csum, singv_lo, shard_idx,
 					     map);
 		if (rc != 0) {
+			bool			 is_ec_obj;
+
 			if (iod->iod_type == DAOS_IOD_SINGLE) {
 				D_ERROR("Data Verification failed (object: "
-					DF_OID"): "DF_RC"\n",
-					DP_OID(orw->orw_oid.id_pub),
+					DF_OID" shard %d): "DF_RC"\n",
+					DP_OID(orw->orw_oid.id_pub), shard_idx,
 					DP_RC(rc));
 			} else  if (iod->iod_type == DAOS_IOD_ARRAY) {
 				D_ERROR("Data Verification failed (object: "
-						DF_OID" , extent: "DF_RECX"):"
-						" "DF_RC"\n",
+					DF_OID" shard %d, extent: "DF_RECX"):"
+					" "DF_RC"\n",
 					DP_OID(orw->orw_oid.id_pub),
-					DP_RECX(iod->iod_recxs[i]),
+					shard_idx, DP_RECX(iod->iod_recxs[i]),
 					DP_RC(rc));
 			}
 
+			is_ec_obj = (reasb_req != NULL) &&
+				DAOS_OC_IS_EC(reasb_req->orr_oca);
+			if (rc == -DER_CSUM && is_ec_obj) {
+				struct shard_auxi_args	*sa;
+				uint32_t		 tgt_idx;
+
+				sa = &rw_args->shard_args->auxi;
+				tgt_idx = sa->shard - sa->start_shard;
+				rc = obj_ec_get_degrade(reasb_req, tgt_idx,
+							NULL, false);
+				if (rc) {
+					D_ERROR(DF_OID" obj_ec_get_degrade "
+						DF_RC"\n",
+						DP_OID(orw->orw_oid.id_pub),
+						DP_RC(rc));
```

Comment:
If rc == -DER_DATA_LOSS, should the return value is still DER_CSUM? which will still cause retry?

Reply:
when there is no more available tgt, the retry will fail immediately.
for that case, user expected to get ECSUM error code.

Reply:
ah, I see. thanks.